features/nanostack/mbed-mesh-api/source/WisunInterface.cpp

```diff
@@ -0,0 +1,185 @@
+{
+    if (!_interface) {
+        _interface = new (nothrow) Nanostack::WisunInterface(*_phy);
+        if (!_interface) {
+            return NSAPI_ERROR_NO_MEMORY;
+        }
+        _interface->attach(_connection_status_cb);
+    }
+    return NSAPI_ERROR_OK;
+}
+
+nsapi_error_t Nanostack::WisunInterface::bringup(bool dhcp, const char *ip,
+                                                 const char *netmask, const char *gw,
+                                                 nsapi_ip_stack_t stack, bool blocking)
+{
+    nanostack_lock();
+
+    if (register_phy() < 0) {
+        nanostack_unlock();
+        return NSAPI_ERROR_DEVICE_ERROR;
+    }
+
+    _blocking = blocking;
+
+    // After the RF is up, we can seed the random from it.
+    randLIB_seed_random();
+
+    mesh_error_t status = init();
+    if (status != MESH_ERROR_NONE) {
+        nanostack_unlock();
+        return map_mesh_error(status);
+    }
+
+    status = mesh_connect();
+    if (status != MESH_ERROR_NONE) {
+        nanostack_unlock();
+        return map_mesh_error(status);
+    }
+
+    // Release mutex before blocking
+    nanostack_unlock();
+
+    if (blocking) {
+        // wait connection for ever
+        int32_t count = connect_semaphore.wait(osWaitForever);
+
+        if (count <= 0) {
+            return NSAPI_ERROR_DHCP_FAILURE; // sort of...
+        }
+    }
+    return 0;
+
+}
+
+nsapi_error_t Nanostack::WisunInterface::bringdown()
+{
+    NanostackLockGuard lock;
+
+    mesh_error_t status = mesh_disconnect();
+
+    return map_mesh_error(status);
+}
+
+mesh_error_t Nanostack::WisunInterface::init()
+{
+    wisun_tasklet_init();
+    __mesh_handler_set_callback(this);
+    interface_id = wisun_tasklet_network_init(_device_id);
+
+    if (interface_id == -2) {
+        return MESH_ERROR_PARAM;
+    } else if (interface_id == -3) {
+        return MESH_ERROR_MEMORY;
+    } else if (interface_id < 0) {
+        return MESH_ERROR_UNKNOWN;
+    }
+    return MESH_ERROR_NONE;
+}
+
+mesh_error_t Nanostack::WisunInterface::mesh_connect()
+{
+    int8_t status = -9; // init to unknown error
+    tr_debug("connect()");
+
+    status = wisun_tasklet_connect(&__mesh_handler_c_callback, interface_id);
+
+    if (status >= 0) {
+        return MESH_ERROR_NONE;
+    } else if (status == -1) {
+        return MESH_ERROR_PARAM;
+    } else if (status == -2) {
+        return MESH_ERROR_MEMORY;
+    } else if (status == -3) {
+        return MESH_ERROR_STATE;
+    } else {
+        return MESH_ERROR_UNKNOWN;
+    }
+}
+
+mesh_error_t Nanostack::WisunInterface::mesh_disconnect()
+{
+    int8_t status = -1;
+
+    status = wisun_tasklet_disconnect(true);
+
+    if (status >= 0) {
+        return MESH_ERROR_NONE;
+    }
+
+    return MESH_ERROR_UNKNOWN;
+}
+
+char *Nanostack::WisunInterface::get_gateway(char *buf, nsapi_size_t buflen)
+{
+    NanostackLockGuard lock;
+    if (wisun_tasklet_get_router_ip_address(buf, buflen) == 0) {
```

Comment:
get_gateway is calling wisun_tasklet_get_router_ip_address and getRouterIpAddress is calling get_gateway?

Reply:
This is implemented just like LOWPANNDInterface.
and WisunInterface::getRouterIpAddress is a api for wisun